### DISCUSSION TITLE
alternateExchange

### DISCUSSION THREAD

**Discussion by paxinos** [rabbitmq/rabbitmq-objc-client#191] [General]
How does one specify an exchange has an alternateExchange?   The methods exposed in this library do not seem to include any ability to specify an alternateExchange.     

For some reason, the Java/Android jar does not cause this fault - even those the alternateExchange is not specified (?)

When attempting to connect to an exchange, rabbitmq's log shows:

2022-03-25 09:58:21.792136-04:00 [erro] <0.15216.8> operation exchange.declare caused a channel exception precondition_failed: inequivalent arg 'alternate-exchange' for exchange 'messages' in vhost 'ezlinq': received none but current is the value 'alternate-messages' of type 'longstr'

For reference, In node-red, the topology is described as below.   WIthout the alternateExchnage the same error (above) is produced.

        {"name": "messages",   "type": "topic", "options": {"durable": true,  "alternateExchange": "alternate-messages"}}

**Comment by michaelklishin** [comment]
Apparently there is no `RMQExchangeDeclareOptions` (an exchange counterpart of `RMQQueueDeclareOptions`). I will file an issue.

**Reply by michaelklishin** [reply]
I'm wrong, the options class does exist but it's not exposed in the API.

**Comment by michaelklishin** [comment]
Just like with optional queue arguments, a much better idea would be to [use a policy](https://rabbitmq.com/ae.html#define). See [why policies exist and how to use them](https://rabbitmq.com/parameters.html#policies). Hardcoding such values in application code
is not recommended.

**Comment by michaelklishin** [comment]
Apparently it is possible to pass in both options and optional arguments, somehow GitHub search failed me :(

From `ExchangeIntegrationTest.swift` (modified a tiny bit):

``` swift
let conn = RMQConnection()
conn.start()
let ch = conn.createChannel()
let x = ch.fanout("objc.tests.fanout", options: [], arguments: [])
```

From `RMQMethods.h`:

``` objc
@interface RMQExchangeDeclare : RMQValue <RMQMethod>
@property (nonnull, copy, nonatomic, readonly) RMQShort *reserved1;
@property (nonnull, copy, nonatomic, readonly) RMQShortstr *exchange;
@property (nonnull, copy, nonatomic, readonly) RMQShortstr *type;
@property (nonatomic, readonly) RMQExchangeDeclareOptions options;
@property (nonnull, copy, nonatomic, readonly) RMQTable *arguments;
- (nonnull instancetype)initWithReserved1:(nonnull RMQShort *)reserved1
                                 exchange:(nonnull RMQShortstr *)exchange
                                     type:(nonnull RMQShortstr *)type
                                  options:(RMQExchangeDeclareOptions)options
                                arguments:(nonnull RMQTable *)arguments;
@end
```

**Reply by michaelklishin** [reply]
So you need to pass an `RMQTable` for `arguments` that matches the map you find in some other clients. Or, again, just use a policy. It's a better idea.
